File: backend/internal/models/expert_subscription.go
package models

import (
	"database/sql"
	"time"
)

// =============================================================================
// ExpertSubscription — one row per (user, expert, attempt). Backed by
// expert_subscriptions (migration 0007).
//
// Lifecycle:
//   pending → active → expired
//          ↘ rejected
//          ↘ cancelled
// =============================================================================

const (
	SubPlanMonthly = "monthly"
	SubPlanYearly  = "yearly"

	SubStatusPending   = "pending"
	SubStatusActive    = "active"
	SubStatusRejected  = "rejected"
	SubStatusCancelled = "cancelled"
	SubStatusExpired   = "expired"

	PaymentMethodCash      = "cash"
	PaymentMethodFIB       = "fib"
	PaymentMethodAppleIAP  = "apple_iap"  // Apple StoreKit2 verified receipt
	PaymentMethodGoogleIAP = "google_iap" // Google Play Billing (future)
)

// Aliases — historical handlers reference these without the Sub prefix.
// Kept so the IAP handler reads cleanly: PlanMonthly / PlanYearly.
const (
	PlanMonthly = SubPlanMonthly
	PlanYearly  = SubPlanYearly
)

// Default prices (USD cents). Plans are server-decided so a user can't ask
// to subscribe at a discount they invented.
const (
	PriceMonthlyCents = 1000 // $10
	PriceYearlyCents  = 9600 // $96 (20% off vs monthly × 12)
)

// PriceForPlan returns the canonical USD-cent price for a plan, or 0 if
// the plan is invalid. Use this in handlers — never trust client prices.
func PriceForPlan(plan string) int {
	switch plan {
	case SubPlanMonthly:
		return PriceMonthlyCents
	case SubPlanYearly:
		return PriceYearlyCents
	}
	return 0
}

// DurationForPlan returns the active-period length for a plan.
func DurationForPlan(plan string) time.Duration {
	switch plan {
	case SubPlanMonthly:
		return 30 * 24 * time.Hour
	case SubPlanYearly:
		return 365 * 24 * time.Hour
	}
	return 0
}

type ExpertSubscription struct {
	ID              int64      `json:"id"`
	UserID          int64      `json:"userId"`
	ExpertID        string     `json:"expertId"`
	Plan            string     `json:"plan"`
	Status          string     `json:"status"`
	PriceCents      int        `json:"priceCents"`
	Currency        string     `json:"currency"`
	PaymentMethod   string     `json:"paymentMethod"`
	PaymentRef      *string    `json:"paymentRef,omitempty"`
	// Cash-receipt image (mig 0024). Optional URL to /uploads/images/...
	// uploaded by the subscriber so the admin can verify the payment.
	ReceiptURL      *string    `json:"receiptUrl,omitempty"`
	UserNote        *string    `json:"userNote,omitempty"`
	RejectionReason *string    `json:"rejectionReason,omitempty"`
	CreatedAt       time.Time  `json:"createdAt"`
	AcceptedAt      *time.Time `json:"acceptedAt,omitempty"`
	AcceptedBy      *int64     `json:"acceptedBy,omitempty"`
	RejectedAt      *time.Time `json:"rejectedAt,omitempty"`
	CancelledAt     *time.Time `json:"cancelledAt,omitempty"`
	ExpiresAt       *time.Time `json:"expiresAt,omitempty"`

	// Joined fields populated when listing for admin dashboard.
	UserEmail     string `json:"userEmail,omitempty"`
	UserName      string `json:"userName,omitempty"`
	ExpertName    string `json:"expertName,omitempty"`
	ReviewerEmail string `json:"reviewerEmail,omitempty"`
}

// IsActiveNow returns true if the subscription is in 'active' state and not
// past its expiry. The DB doesn't auto-flip to 'expired' so callers must
// also do this check at read time.
func (s *ExpertSubscription) IsActiveNow() bool {
	if s == nil || s.Status != SubStatusActive {
		return false
	}
	if s.ExpiresAt == nil {
		return false
	}
	return s.ExpiresAt.After(time.Now())
}

type subRowScanner interface {
	Scan(dest ...any) error
}

// ScanSubscription scans columns in this exact order:
//
//	id, user_id, expert_id, plan, status, price_cents, currency,
//	payment_method, payment_ref, user_note, rejection_reason,
//	created_at, accepted_at, accepted_by, rejected_at, cancelled_at, expires_at,
//	receipt_url
func ScanSubscription(s subRowScanner) (*ExpertSubscription, error) {
	var sub ExpertSubscription
	var (
		paymentRef      sql.NullString
		userNote        sql.NullString
		rejectionReason sql.NullString
		acceptedAt      sql.NullTime
		acceptedBy      sql.NullInt64
		rejectedAt      sql.NullTime
		cancelledAt     sql.NullTime
		expiresAt       sql.NullTime
		receiptURL      sql.NullString
	)
	if err := s.Scan(
		&sub.ID, &sub.UserID, &sub.ExpertID, &sub.Plan, &sub.Status,
		&sub.PriceCents, &sub.Currency,
		&sub.PaymentMethod, &paymentRef, &userNote, &rejectionReason,
		&sub.CreatedAt, &acceptedAt, &acceptedBy, &rejectedAt, &cancelledAt, &expiresAt,
		&receiptURL,
	); err != nil {
		return nil, err
	}
	if paymentRef.Valid {
		sub.PaymentRef = &paymentRef.String
	}
	if userNote.Valid {
		sub.UserNote = &userNote.String
	}
	if rejectionReason.Valid {
		sub.RejectionReason = &rejectionReason.String
	}
	if acceptedAt.Valid {
		sub.AcceptedAt = &acceptedAt.Time
	}
	if acceptedBy.Valid {
		sub.AcceptedBy = &acceptedBy.Int64
	}
	if rejectedAt.Valid {
		sub.RejectedAt = &rejectedAt.Time
	}
	if cancelledAt.Valid {
		sub.CancelledAt = &cancelledAt.Time
	}
	if expiresAt.Valid {
		sub.ExpiresAt = &expiresAt.Time
	}
	if receiptURL.Valid {
		sub.ReceiptURL = &receiptURL.String
	}
	return &sub, nil
}

File: lib/models/expert_subscription.dart
/// Plan choice — fixed by the server. Monthly = $10, Yearly = $96.
enum SubPlan { monthly, yearly }

extension SubPlanX on SubPlan {
  String get wireValue {
    switch (this) {
      case SubPlan.monthly: return 'monthly';
      case SubPlan.yearly:  return 'yearly';
    }
  }

  String get label {
    switch (this) {
      case SubPlan.monthly: return 'Monthly';
      case SubPlan.yearly:  return 'Yearly';
    }
  }

  /// Canonical price in USD cents. Mirrors backend `models.PriceForPlan`.
  int get priceCents {
    switch (this) {
      case SubPlan.monthly: return 1000;
      case SubPlan.yearly:  return 9600;
    }
  }

  /// "$10" / "$96"
  String get priceLabel => '\$${(priceCents / 100).toStringAsFixed(0)}';

  /// Days the subscription stays active after acceptance.
  int get durationDays {
    switch (this) {
      case SubPlan.monthly: return 30;
      case SubPlan.yearly:  return 365;
    }
  }

  static SubPlan fromWire(String? raw) {
    switch ((raw ?? 'monthly').toLowerCase()) {
      case 'yearly': return SubPlan.yearly;
      default:       return SubPlan.monthly;
    }
  }
}

/// Lifecycle of a subscription request.
enum SubStatus { pending, active, rejected, cancelled, expired }

extension SubStatusX on SubStatus {
  String get wireValue {
    switch (this) {
      case SubStatus.pending:   return 'pending';
      case SubStatus.active:    return 'active';
      case SubStatus.rejected:  return 'rejected';
      case SubStatus.cancelled: return 'cancelled';
      case SubStatus.expired:   return 'expired';
    }
  }

  static SubStatus fromWire(String? raw) {
    switch ((raw ?? 'pending').toLowerCase()) {
      case 'active':    return SubStatus.active;
      case 'rejected':  return SubStatus.rejected;
      case 'cancelled': return SubStatus.cancelled;
      case 'expired':   return SubStatus.expired;
      default:          return SubStatus.pending;
    }
  }
}

/// Payment method picked by the user when submitting a subscription.
///   * cash       — pay admin in person
///   * fib        — First Iraqi Bank transfer (paymentRef = transaction id)
///   * appleIap   — Apple StoreKit (auto-verified via /me/iap/apple/verify)
///   * googleIap  — Google Play Billing (Phase 2.4+, not yet wired here)
enum PaymentMethod { cash, fib, appleIap, googleIap }

extension PaymentMethodX on PaymentMethod {
  String get wireValue {
    switch (this) {
      case PaymentMethod.cash:      return 'cash';
      case PaymentMethod.fib:       return 'fib';
      case PaymentMethod.appleIap:  return 'apple_iap';
      case PaymentMethod.googleIap: return 'google_iap';
    }
  }

  String get label {
    switch (this) {
      case PaymentMethod.cash:      return 'Cash';
      case PaymentMethod.fib:       return 'FIB';
      case PaymentMethod.appleIap:  return 'Apple Pay';
      case PaymentMethod.googleIap: return 'Google Play';
    }
  }

  static PaymentMethod fromWire(String? raw) {
    switch ((raw ?? 'cash').toLowerCase()) {
      case 'fib':        return PaymentMethod.fib;
      case 'apple_iap':  return PaymentMethod.appleIap;
      case 'google_iap': return PaymentMethod.googleIap;
      default:           return PaymentMethod.cash;
    }
  }
}

/// One row from `expert_subscriptions`.
class ExpertSubscription {
  final int id;
  final int userId;
  final String expertId;
  final SubPlan plan;
  final SubStatus status;
  final int priceCents;
  final String currency;
  final PaymentMethod paymentMethod;
  final String? paymentRef;
  /// Cash-receipt image URL (mig 0024). Null when not uploaded.
  final String? receiptUrl;
  final String? userNote;
  final String? rejectionReason;
  final DateTime createdAt;
  final DateTime? acceptedAt;
  final int? acceptedBy;
  final DateTime? rejectedAt;
  final DateTime? cancelledAt;
  final DateTime? expiresAt;

  // Optional joined fields surfaced by the admin list endpoint.
  final String? userEmail;
  final String? userName;
  final String? expertName;

  const ExpertSubscription({
    required this.id,
    required this.userId,
    required this.expertId,
    required this.plan,
    required this.status,
    required this.priceCents,
    required this.currency,
    required this.paymentMethod,
    required this.createdAt,
    this.paymentRef,
    this.receiptUrl,
    this.userNote,
    this.rejectionReason,
    this.acceptedAt,
    this.acceptedBy,
    this.rejectedAt,
    this.cancelledAt,
    this.expiresAt,
    this.userEmail,
    this.userName,
    this.expertName,
  });

  /// True when this subscription is currently active (status + not expired).
  bool get isActiveNow {
    if (status != SubStatus.active) return false;
    final exp = expiresAt;
    if (exp == null) return false;
    return exp.isAfter(DateTime.now());
  }

  /// Days remaining until expiry — null when not active.
  int? get daysRemaining {
    if (!isActiveNow) return null;
    return expiresAt!.difference(DateTime.now()).inDays;
  }

  /// "$10.00" — formatted from priceCents.
  String get priceLabel {
    final dollars = priceCents / 100;
    return '\$${dollars.toStringAsFixed(dollars.truncateToDouble() == dollars ? 0 : 2)}';
  }

  factory ExpertSubscription.fromJson(Map<String, dynamic> json) =>
      ExpertSubscription(
        id: (json['id'] as num).toInt(),
        userId: (json['userId'] as num).toInt(),
        expertId: json['expertId'] as String? ?? '',
        plan: SubPlanX.fromWire(json['plan'] as String?),
        status: SubStatusX.fromWire(json['status'] as String?),
        priceCents: (json['priceCents'] as num?)?.toInt() ?? 0,
        currency: json['currency'] as String? ?? 'USD',
        paymentMethod: PaymentMethodX.fromWire(json['paymentMethod'] as String?),
        paymentRef: json['paymentRef'] as String?,
        receiptUrl: json['receiptUrl'] as String?,
        userNote: json['userNote'] as String?,
        rejectionReason: json['rejectionReason'] as String?,
        createdAt: DateTime.tryParse(json['createdAt'] as String? ?? '') ??
            DateTime.now(),
        acceptedAt: _parseDate(json['acceptedAt']),
        acceptedBy: (json['acceptedBy'] as num?)?.toInt(),
        rejectedAt: _parseDate(json['rejectedAt']),
        cancelledAt: _parseDate(json['cancelledAt']),
        expiresAt: _parseDate(json['expiresAt']),
        userEmail: json['userEmail'] as String?,
        userName: json['userName'] as String?,
        expertName: json['expertName'] as String?,
      );

  static DateTime? _parseDate(dynamic v) {
    if (v == null) return null;
    if (v is String && v.isEmpty) return null;
    return DateTime.tryParse(v.toString());
  }
}
